api/v1beta1/module_webhook.go

```
@@ -0,0 +1,108 @@
+)
+
+// log is for logging in this package.
+var modulelog = logf.Log.WithName("module-resource")
+
+func (r *Module) SetupWebhookWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(r).
+		Complete()
+}
+
+//+kubebuilder:webhook:path=/validate-kmm-sigs-x-k8s-io-v1beta1-module,mutating=false,failurePolicy=fail,sideEffects=None,groups=kmm.sigs.x-k8s.io,resources=modules,verbs=create;update,versions=v1beta1,name=vmodule.kb.io,admissionReviewVersions=v1
+
+var _ webhook.Validator = &Module{}
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type
+func (r *Module) ValidateCreate() error {
+	modulelog.Info("validate create", "name", r.Name)
+
+	if err := r.validateKernelMappingRegex(); err != nil {
+		return err
+	}
+
+	return r.validateModuleName()
+}
+
+// ValidateUpdate implements webhook.Validator so a webhook will be registered for the type
+func (r *Module) ValidateUpdate(_ runtime.Object) error {
+	modulelog.Info("validate update", "name", r.Name)
+
+	if err := r.validateKernelMappingRegex(); err != nil {
+		return err
+	}
+
+	return r.validateModuleName()
+}
+
+// ValidateDelete implements webhook.Validator so a webhook will be registered for the type
+func (r *Module) ValidateDelete() error {
+	return nil
+}
+
+func (r *Module) validateKernelMappingRegex() error {
+	if r.Spec.ModuleLoader.Container.KernelMappings == nil {
+		return nil
+	}
+
+	container := r.Spec.ModuleLoader.Container
+
+	for idx, km := range container.KernelMappings {
+		if _, err := regexp.Compile(km.Regexp); err != nil {
+			return fmt.Errorf("invalid regexp at index %d: %v", idx, err)
+		}
+
+		if len(container.ContainerImage) == 0 && len(km.ContainerImage) == 0 {
+			return fmt.Errorf("missing spec.moduleLoader.container.kernelMappings[].containerImage")
+		}
+	}
+
+	return nil
+}
+
+func (r *Module) validateModuleName() error {
+	modprobe := r.Spec.ModuleLoader.Container.Modprobe
+
+	if len(modprobe.ModuleName) > 0 {
```

```suggestion
	if modprobe.ModuleName != "" {
```

Also, what happens if both `modprobe.ModuleName` and `modprobe.RawArgs` are set? We should return an error in this case aren't we?

---

Not sure. I understood that the `modprobe.ModuleName` has a preference and if it's set, we can ignore the `RawArgs`, but maybe I misunderstood the behavior. 

@qbarrand can you confirm this?